File: dydx/dydxTurnkey/dydxTurnkey/TurnkeyNativeModule.m
//
//  TurnkeyNativeModule.m
//  dydxTurnkey
//
//  Created by Rui Huang on 16/07/2025.
//

#import <Foundation/Foundation.h>

#import "React/RCTBridgeModule.h"
#import "React/RCTEventEmitter.h"

@interface RCT_EXTERN_MODULE(TurnkeyNativeModule, RCTEventEmitter)

/**
  `RCT_EXTERN_METHOD` in React Native allows exposing native methods to the JavaScript bridge.
 
  It follows the syntax: `RCT_EXTERN_METHOD(methodName:(paramType)internalParamName ...)`
  
  For one argument: `RCT_EXTERN_METHOD(methodName:(ParamType)internalParamName)`
 
  For multiple arguments: `RCT_EXTERN_METHOD(methodName:(ParamType1)internalParamName1 [externalParamName2]:(ParamType2)internalParamName2 ...)`
 */
RCT_EXTERN_METHOD(onJsResponse
                  :(NSString) callbackId
                  :(NSString) result)

RCT_EXTERN_METHOD(onAuthRouteToWallet)
RCT_EXTERN_METHOD(onAuthRouteToDesktopQR)
RCT_EXTERN_METHOD(onAuthCompleted
                  :(NSString) onboardingSignature
                  :(NSString) evmAddress
                  :(NSString) svmAddress
                  :(NSString) mnemonics
                  :(NSString) loginMethod
                  :(NSString) userEmail
                  :(NSString) dydxAddress)

RCT_EXTERN_METHOD(onAppleAuthRequest
                  :(NSString) nonce)

RCT_EXTERN_METHOD(onTrackingEvent
                  :(NSString) eventName
                  :(NSDictionary *)eventParams)

@end

File: dydx/dydxTurnkey/dydxTurnkey/TurnkeyNativeModule.swift
//
//  TurnkeyNativeModule.swift
//  dydxTurnkey
//
//  Created by Rui Huang on 15/07/2025.
//

import React
import Foundation
internal import Utilities

@objc(TurnkeyNativeModule)
class TurnkeyNativeModule: NSObject, RCTBridgeModule {
    static func moduleName() -> String {
        return "TurnkeyNativeModule"
    }

    static func requiresMainQueueSetup() -> Bool {
      return false
    }

    weak var delegate: TurnkeyBridgeManagerDelegate?

    private var pendingCompletions: [String: (String) -> Void] = [:]

    func callMyJsFunction(functionName: String,
                          params: [String: String] = [:],
                          completion: @escaping (String) -> Void) {
        let bridge = TurnkeyBridgeManager.shared.bridge
        let callbackId = UUID().uuidString

        // Store completion for callback correlation
        pendingCompletions[callbackId] = completion

        var params = params
        params["callbackId"] = callbackId

        bridge.enqueueJSCall(
          "RCTDeviceEventEmitter",
          method: "emit",
          args: [functionName, params],
          completion: nil
        )
    }

    @objc(onJsResponse::)
    func onJsResponse(_ callbackId: String, _ result: String) {
        if let completion = pendingCompletions[callbackId] {
            DispatchQueue.main.async {
                completion(result)
            }
            pendingCompletions.removeValue(forKey: callbackId)
        }
    }

    @objc(onAuthRouteToWallet)
    func onAuthRouteToWallet() {
        DispatchQueue.main.async { [weak self] in
            self?.delegate?.onAuthRouteToWallet()
        }
    }

    @objc(onAuthRouteToDesktopQR)
    func onAuthRouteToDesktopQR() {
        DispatchQueue.main.async { [weak self] in
            self?.delegate?.onAuthRouteToDesktopQR()
        }
    }

    @objc(onAuthCompleted:::::::)
    func onAuthCompleted(onboardingSignature: String, evmAddress: String, svmAddress: String, mnemonics: String, loginMethod: String, userEmail: String?, dydxAddress: String?) {
        DispatchQueue.main.async { [weak self] in
            self?.delegate?.onAuthCompleted(onboardingSignature: onboardingSignature, evmAddress: evmAddress, svmAddress: svmAddress, mnemonics: mnemonics, loginMethod: loginMethod, userEmail: userEmail, dydxAddress: dydxAddress)
        }
    }

    @objc(onAppleAuthRequest:)
    func onAppleAuthRequest(nonce: String) {
        DispatchQueue.main.async { [weak self] in
            self?.delegate?.onAppleAuthRequest(nonce: nonce)
        }
    }

    @objc(onTrackingEvent::)
    func onTrackingEvent(eventName: String, eventParams: [String: String]) {
        Tracking.shared?.log(event: eventName, data: eventParams)
    }
}
